distributed/scheduler.py

```diff
@@ -2074,7 +2075,16 @@ def decide_worker_rootish_queuing_disabled(
 
         return ws
 
-    def decide_worker_rootish_queuing_enabled(self) -> WorkerState | None:
+    def worker_objective_rootish_queuing(self, ws, ts):
+        # FIXME: This is basically the ordinary worker_objective but with task
+        # counts instead of occupancy.
+        comm_bytes = sum(
+            dts.get_nbytes() for dts in ts.dependencies if ws not in dts.who_has
+        )
+        # See test_nbytes_determines_worker
+        return (len(ws.processing) / ws.nthreads, comm_bytes, ws.nbytes)
```

Comment:
@gjoseph92 I remember you very strongly opposing adding an objective function that respect comm_bytes. I remember vaguely a point about widely shared dependencies but I do not see an issue with this.

We need an objective like this to ensure very basic scheduling policies to be respected, e.g. `test_nbytes_determines_worker` where a dependent task is supposed to be scheduled on a worker holding most of its dependencies.

This should only truly matter if there are multiple workers available with a similar number of free slots. My intuition tells me this objective function should still make sense

Reply:
xref https://github.com/dask/distributed/pull/7280

Reply:
> `test_nbytes_determines_worker` where a dependent task is supposed to be scheduled on a worker holding most of its dependencies

**tl;dr** then that task shouldn't have been queued in the first place, and the root-ish heuristic is at fault, not the objective function. For a task to be queued, it must be equally good to run on any worker.

I do oppose an objective function that considers data transfer _for root-ish tasks_. It contradicts the definition of root-ish-ness: that the location of the tasks' dependencies are not important for scheduling purposes.

When there's only one thread open in the cluster, on a random worker, the next task on the queue must be equally "good" to run on that random worker as on any other worker in the cluster. If it would have been "better" if the thread had opened on a different particular worker—because that worker had more dependencies in memory, say, or because it had some resource—then the task should not have been on the queue in the first place. (This is why putting tasks with restrictions on the queue breaks the core assumption of queuing: https://github.com/dask/distributed/pull/7526.)

This is important because we choose the pool of candidate workers for queued tasks based only on how busy they are. That works because we can guarantee there are no other factors that should play into worker selection for these tasks.

The root-ish-ness heuristic is meant to provide that guarantee, by identifying task groups where:
1. The tasks will end up being assigned to every worker in the cluster (`len > nthreads`)
2. Therefore, any common dependencies the tasks have will end up being replicated onto every worker in the cluster (`deps len < 5`)

The key is the second point: because the dependencies will be replicated to all workers, we don't need to select workers that avoid data transfer—in the long term, the data transfer is unavoidable. Indeed, we even know of some cases (that don't meet the root-ish heuristic, to be clear) where trying to avoid a transfer leads to worse scheduling choices: https://github.com/dask/distributed/issues/6570.

So the reason I oppose an objective function that considers data transfer for root-ish tasks is because it _implies_ that one worker might be preferable over another based on something besides busyness, which violates these core assumptions.

But to be clear: I only oppose this on the grounds of logical consistency. I don't expect an objective function that considers transfer will actually make a measurable different in most cases. (In most cases, the size of the candidate pool is 1 anyway.) But since it contradicts this key assumption of queuing, I'd want to come up with logical proof of why it's safe—and why it's preferable—especially since we know this type of weighting leads to bad decisions in other cases (https://github.com/dask/distributed/issues/6570).

I haven't bothered to put in the effort to make the proof, since I haven't seen what the gain would be. I'm happy to be shown that it's both safe and better, though—again, I just haven't seen a reason to.

See also:
- https://github.com/dask/distributed/pull/7280#discussion_r1018608580
- https://github.com/dask/distributed/pull/7280#discussion_r1019420197

Reply:
First of all, I think this is a moot point once I introduce some cutoff as suggested in the other comment. It would've been fine to just point out this error instead of writing such a long comment.

> It contradicts the definition of root-ish-ness: that the location of the tasks' dependencies are not important for scheduling purposes.

I disagree with this definition. The reason why we're having a special case for these tasks is because we assume they are data producers. The natural assumption is that they are at the bottom of our graph and are the most frequent source of data overproduction.
While I see some value for intermediate layers to be considered as root-ish I believe the same argument holds.
I don't think these perspectives are orthogonal but your definition is not the whole truth.

If something like https://github.com/dask/distributed/issues/6570 is truly the complexity driver here we should consider alternative approaches (inlining, better weights, scheduler hints, etc.)


> If it would have been "better" if the thread had opened on a different particular worker—because that worker had more dependencies in memory, say, or because it had some resource—then the task should not have been on the queue in the first place.

I also don't agree with this sentiment. I see a point for suppressing/ignoring dependencies but especially for cases where worker-saturation>1 I see how nuance can matter. Total randomness is not necessarily desired.